New Keysanity Variants: Overworld and Dungeons

Cuphat opened the pull request:
A new variant of keysanity which allows Maps & Compasses, Small Keys, Boss Keys, and Ganon's Boss Key to be restricted to the overworld or inside of any dungeon in addition to anywhere.

Additionally: Split Gerudo Fortress Small Keys to separate setting. Gerudo Fortress Keys have always been the red-headed stepchild of the small key setting. This splits it off to allow more control over these keys.

Cuphat committed:
New Keysanity Variants: Overworld and Dungeons

A new variant of keysanity which allows Maps & Compasses, Small Keys, Boss Keys, and Ganon's Boss Key to be restricted to the overworld or inside of any dungeon in addition to anywhere.

fenhl commented on SettingsList.py:
```diff
@@ -2536,7 +2546,18 @@ def __init__(self, name, gui_text, min, max, default, step=1,
+            
+            'Overworld Only': Small Keys can only appear outside
+            of dungeons. You may need to enter a dungeon multiple
+            times to gain items to access to the overworld locations
```
```suggestion
            times to gain items to access the overworld locations
```

fenhl commented on SettingsList.py:
```diff
@@ -2611,6 +2647,12 @@ def __init__(self, name, gui_text, min, max, default, step=1,
             'Vanilla': Ganon's Castle Boss Key will appear in 
             the vanilla location.
+            
+            'Overworld Only': Ganon's Castle Boss Key can only appear
```
```suggestion
            'Overworld Only': Ganon's Castle Boss Key can only
```

Cuphat commented:
Resolved the typos above and force-pushed. Thanks, @fenhl for catching those.

r0bd0g commented:
At this point I feel like people are adding options just to screw with rando rando players lol.

This PR puts thieves' hideout into an awkward spot. It's neither a dungeon nor overworld? Then where is it?

Cuphat commented:
Thieves' Hideout, to rando, is overworld. Whether it should be considered a dungeon or not seems like a philosophical question, especially since the game itself treats it as one in a lot of ways.

As for the fortress keys themselves, they are in a pretty weird spot all around and I am not sure the best way to handle those.

Cuphat committed:
Split Gerudo Fortress Small Keys to separate setting.

Gerudo Fortress Keys have always been the red-headed stepchild of the small key setting. This splits it off to allow more control over these keys.

puddydoo commented:
I think that Thieves' Hideout should be treated as overworld, but if the hideout keys are set to dungeons only, then it seems arbitrary to make them appear only in other dungeons. Dungeon-only fortress key setting should treat the 4 key locations and Gerudo Card as both dungeon and overworld for purposes of keysanity so that other "any dungeon" keys can appear there.

r0bd0g commented:
The dungeons setting is incompatible with 1 major item per dungeon, btw. Boss keys only in dungeons might gen but you'd probably have to make sure they got placed first so that there'd always be room.

Cuphat committed:
Prioritize dungeon restricted items with 'Dungeons Have One Major Item'

Dungeon restricted items are now placed first when placing items for 'Dungeons Have One Major Item'. Setting `Shuffle Small Keys` to `Any Dungeon` now disables `Dungeons Have One Major Item` since there are far too many keys for these settings to cooperate.

Some other minor cleanup regarding key distinctions. Renamed the options for `Shuffle Gerudo Fortress Keys` to be consistently named with the other shuffle key settings.

Cuphat commented:
I added in some handling for `Dungeons Only Have One Major Item` which is a setting I always forget about and dislike. Also did other minor cleanup.

Cuphat committed:
Altar text for Ganon BK Overworld Only/Any Dungeon.

Zannick commented on SettingsList.py:
```diff
@@ -2519,7 +2530,9 @@ def __init__(self, name, gui_text, min, max, default, step=1,
             'vanilla':   'Vanilla Locations',
             'dungeon':   'Dungeon Only',
-            'keysanity': 'Anywhere (Keysanity)'
+            'overworld': 'Overworld Only (Keysanity)',
```
I don't think it makes much sense to say "(Keysanity)" after all of these. I could see that leading to confusion if people say "I'm playing Keysanity." Maybe keep to only the last one, or remove them all? Same for the others.

Cuphat replied:
Done.

Zannick commented on SettingsList.py:
```diff
@@ -2481,7 +2484,9 @@ def __init__(self, name, gui_text, min, max, default, step=1,
             'dungeon':   'Dungeon Only',
-            'keysanity': 'Anywhere'
+            'overworld': 'Overworld Only',
+            'dungeons':  'Any Dungeon',
```
I'd suggest 'anydungeon' for the choice key, or 'mydungeon' for the 'Dungeon Only' option, just to keep the choice keys harder to confuse. Same for all the others.

Cuphat replied:
I went with `any_dungeon` but done.

Cuphat committed:
Rename key for "Any Dungeon" key shuffle options.
